src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs

```diff
@@ -0,0 +1,203 @@
+        {
+            if (_suppressLaunchBrowser)
+            {
+                return default;
+            }
+
+            if (context.Iteration == 0)
+            {
+                _reporter = context.Reporter;
+
+                if (CanLaunchBrowser(context, out var launchPath))
+                {
+                    _canLaunchBrowser = true;
+                    _launchPath = launchPath;
+
+                    _refreshServer = new BrowserRefreshServer(context.Reporter);
+                    var serverUrl = _refreshServer.Start();
+
+                    context.Reporter.Verbose($"Refresh server running at {serverUrl}.");
+                    context.ProcessSpec.EnvironmentVariables["DOTNET_WATCH_REFRESH_URL"] = serverUrl;
+
+                    context.ProcessSpec.OnOutput += OnOutput;
+                }
+            }
+
+            if (_canLaunchBrowser)
+            {
+                if (context.Iteration > 0)
+                {
+                    // We've detected a change. Notify the browser.
+                    _refreshServer.SendMessage(WaitMessage);
+                }
+            }
+
+            return default;
+        }
+
+        private void OnOutput(object sender, DataReceivedEventArgs eventArgs)
+        {
+            // We've redirected the output, but want to ensure that continues to appear in the user's console.
+            Console.WriteLine(eventArgs.Data);
+
+            if (string.IsNullOrEmpty(eventArgs.Data))
+            {
+                return;
+            }
+
+            if (ApplicationStartedRegex.IsMatch(eventArgs.Data))
+            {
+                var process = (Process)sender;
+                process.OutputDataReceived -= OnOutput;
+                process.CancelOutputRead();
+            }
+            else
+            {
+                var match = NowListeningRegex.Match(eventArgs.Data);
+                if (match.Success)
+                {
+                    var launchUrl = match.Groups["url"].Value;
+
+                    if (!_browserLaunched)
+                    {
+                        _browserLaunched = true;
+                        try
+                        {
+                            LaunchBrowser(launchUrl);
+                        }
+                        catch (Exception ex)
+                        {
+                            _reporter.Output($"Unable to launch browser: {ex}");
+                            _canLaunchBrowser = false;
+                        }
+                    }
+                    else
+                    {
+                        _reporter.Verbose($"Reloading browser");
+                        _refreshServer.SendMessage(ReloadMessage);
+                    }
+                }
+            }
+        }
+
+        private void LaunchBrowser(string launchUrl)
+        {
+            var fileName = launchUrl + "/" + _launchPath;
+            var args = string.Empty;
+            if (!string.IsNullOrEmpty(_browserPath))
```

Comment:
Should there be an error here? If you didn't set a browser path it'll try to run "http://localhost:54304/something" as a command which I don't think works